### PR TITLE
[core] Fix predicate literals cast in filter pushdown after schema evolution

### PR DESCRIPTION
**Pull request by yuzelin**
### Purpose

Reproduce case:
```
CREATE TABLE T (
  id INT,
  f DECIMAL(5,2),
);
INSERT INTO T VALUES (1, 111.32);
ALTER TABLE T MODIFY (f DECIMAL(6, 3));
SELECT * FROM T WHERE f < 111.321;
```
The result is empty but expected 111.320

There are also same problem for other field cast Scenario.

### CODE REVIEW

File: paimon-common/src/main/java/org/apache/paimon/casting/CastExecutors.java [hunk 1]
```diff
@@ -100,6 +115,147 @@ public class CastExecutors {
         return IDENTITY_CAST_EXECUTOR;
     }
 
+    /**
+     * When filter a field witch was evolved from/to a numeric type, we should carefully handle the
+     * precision match and overflow problem.
+     */
+    @Nullable
+    public static List<Object> safelyCastLiteralsWithNumericEvolution(
```
Comment (JingsongLi) [inline paimon-common/src/main/java/org/apache/paimon/casting/CastExecutors.java:123]: Optional<List<Object>>

File: paimon-common/src/main/java/org/apache/paimon/casting/CastExecutors.java [hunk 2]
```diff
@@ -100,6 +115,147 @@ public class CastExecutors {
...
+            LeafPredicate predicate, DataType outputType) {
```
Comment (JingsongLi) [inline paimon-common/src/main/java/org/apache/paimon/casting/CastExecutors.java:124]: List<Object> literals, DataType inputType, DataType outputType

File: paimon-common/src/main/java/org/apache/paimon/casting/CastExecutors.java (hunk 1 above)
Comment (JingsongLi) [inline paimon-common/src/main/java/org/apache/paimon/casting/CastExecutors.java:123]: castLiteralsWithNumericEvolution

**Comment by wwj6591812** [comment]
+1

**Comment by BitFunnelComp** [comment]
The predicate conversion of decimal/floating type seems to be too complicated and there may be bad cases. Is it possible to set up a whitelist to only allow integer conversion?

**Comment by yuzelin** [comment]
> The predicate conversion of decimal/floating type seems to be too complicated and there may be bad cases. Is it possible to set up a whitelist to only allow integer conversion?

I think this is a good idea. Actually, decimal type and related schema evolution is rarely used in industry.

**Review by JingsongLi — COMMENTED** [review]
Add a large BIGINT (Bigger than Int.Max) testing.

File: paimon-common/src/main/java/org/apache/paimon/casting/CastExecutors.java [hunk 3]
```diff
@@ -100,6 +102,63 @@ public class CastExecutors {
...
+                        return Optional.of(literals);
+                    }
+
+                    // Parquet filter Int comparator cannot handle long value.
+                    // See org.apache.parquet.schema.PrimitiveType.
+                    // So ignore filter if long literal is out of int scale.
+                    List<Object> newLiterals = new ArrayList<>(literals.size());
+                    for (Object literal : literals) {
+                        long originalValue = (long) literal;
+                        int newValue = (int) originalValue;
```
Comment (JingsongLi) [inline paimon-common/src/main/java/org/apache/paimon/casting/CastExecutors.java:142]: Create a CastExecutor to do this casting?

**Comment by JingsongLi** [comment]
+1

**Pull request merged**